### ISSUE TITLE
Add initial support for Ingress V2 types

### ISSUE THREAD

**Issue by youngnick** [projectcontour/contour#2113]
As part of the work I'm doing with the service-api subproject of SIG-Network, we're building new CRDs which will serve as the definition of what the next version of Ingress needs.

Contour will need to support these, so this issue is to cover figuring out how to wire up the new types into Contour, and put them behind an experimental flag so that we can get this wiring in and update the types as they are worked on.

That will allow us to have a headstart on implementing them as they get closer to being finalised.

Putting the usage of them behind an experimental feature flag should insulate the rest of the project from these concerns, hopefully.

Comments on the approach welcomed.

**Comment by jpeach** [comment]
> Putting the usage of them behind an experimental feature flag should insulate the rest of the project from these concerns

Do you mean a Go build tag? A configuration flag?

**Comment by youngnick** [comment]
A CLI flag, sorry. No config option for now. This is mainly about ensuring that we don't need a long-running feature branch.

**Comment by youngnick** [comment]
To consume the service-apis types in Contour, we first need to be able to import the Go types somehow. There are three options:
- Generate the Informers code for the new types the same way we do for Contour's types.
- Move Contour to use the dynamic client, so we don't need any generated code at all.
- Move Contour to use the same controller-runtime/kubebuilder framework that the service-apis use (#2118)

I spent some time spiking the controller-runtime/kubebuilder framework, and it probably is worth doing at some point, but it's a substantial amount of change, and would require careful design. It's probably too much work for this point in time. The PR with a design was #2120, and I have a branch in my Contour fork with the changes if we want to come back to them. (https://github.com/youngnick/contour/tree/reconcile-style)

I also had a look at generating the Informers code, but it seems that the code-generation tools do some magic around the use of `k8s.io` for package names that make using them for generating the service-apis types difficult. I've created https://github.com/kubernetes-sigs/service-apis/issues/61 in the service-apis repo, and I'll update that with my results shortly.

@stevesloka has done a great job of trying out the dynamic client in #2191, and that looks like a viable option. Steve and I would both like @davecheney to take a look before we go any further though.

Once we have a way to import the types, the next steps are:
- Create the new CLI flag (`--experimental-service-apis`) to enable use of the new alpha types
- Add watching of the new types to Contour behind that flag
- Add placeholder handling of the new types into the eventhandler. To begin with, just logging that we've seen them is okay.

Once we've made a call about the types, I'll create issues covering this work.
